Model, Fixed fineOne incorrect initial order

Yahasana opened the pull request:
initial model data should go ahead `initialize`

Commit:
Model, Fixed fineOne incorrect initial order

initial model data should go ahead `initialize`

Commit:
Model, empty is better performance than count

mruz commented:
But then we will get the `Not found` exception if no result, previously it was `false`.
```php
$u = Users::findOne(['username' => 'abcd']);
```

Commit:
Model, fixed getRules return all rules

$model->getRules(['nofound']) should not get all the rules

Commit:
Model, remove no found exception

only update primary key when insert record has an auto increment primary key.
latest static binding for model instance creation.

Commit:
fixed findOne

Commit:
Fix loadOne([]) will load unexpect one

Commit:
Minus enhance findOne

no need to create instance twice

Yahasana commented:
Model should be fine now. it's there exact release date for v1.4.0 ? @mruz

mruz commented:
I'll add some tests for the model and merge after. Just to be sure it's working fine. 
There is no strict release date atm.

Yahasana commented:
Initial flow test case
```php
class TestModel extends Model
{
    public function __construct($filters = null, array $data = [])
    {
        echo 'model contruct begin';
        parent::__construct($filters, $data);
        echo 'model contruct done';
        print_r($this->all());
    }
    public function onContruct()
    {
        echo 'model contructing...';
        print_r($this->all());
        echo 'model set columnName=onContruct';
        $this->set('columnName', 'onContruct'); // real columnName
    }
    public function initialize()
    {
        echo 'model initialize...';
        print_r($this->all());
        echo 'model set columnName=initialize';
        $this->set('columnName', 'initialize'); // real columnName
    }
}
```

mruz commented:
I pushed few db/model tests. Could you please rebase your branch?
```
git remote add upstream https://github.com/ice/framework.git
git fetch upstream
git rebase upstream/dev
git push origin patch-10
```

Yahasana commented:
it'd be bug of unitest. after some investigations, you can't write unitest case for `static` things like as `non-static` one

mruz commented on ice/mvc/model.zep:
```diff
@@ -147,14 +151,16 @@ abstract class Model extends Arr implements \Serializable
 
         if result->count() {
             for data in iterator(result) {
-                let instances[] = create_instance_params(get_called_class(), [null, data]);
```
As I remember I had some issues with static implementation on zephir and that's why I used
```zep
create_instance_params(get_called_class(), [null, data]);
```

mruz commented on ice/mvc/model.zep:
```diff
@@ -178,13 +184,9 @@ abstract class Model extends Arr implements \Serializable
-
-        return result;
+        var model;
+        let model = new static(filters, options);
```
Options are not the same as data in construct.

Options are passed to the db::find query
```
// eg. find the latest active user
$user = User::findOne(['status' => 1], ['order' => 'id DESC']);
```

Data is used to populate the fields
```
// fills the $user->data attribute
$user = new User(null, ['username' => 'ice', 'status' => 1]);
```

mruz reviewed (changes requested):
It would be good to continue in #183

Yahasana commented:
new [pr](https://github.com/ice/framework/pull/185)
